doc/user/interop.md

```diff
@@ -89,150 +89,109 @@ Returns true only for the `nil` object.
 
-`READABLE` will be set if there is an instance variable of that name.
+- `READABLE` will be set if the object has a method of that name, or the object
+  has a `[]` method.
 
-`MODIFIABLE` will be set if there is an instance variable of that name and the
-object is not frozen.
+- `INSERTABLE` will be set if the object has a `[]=` method, or the name is a
+  legal instance variable name.
 
-`INSERTABLE` will be set if the object is not frozen.
+- `MODIFIABLE` will be set if the object has a `[]=` method, or the name is an
+  existing instance variable.
 
-`INTERNAL` will be set if there is an instance variable of that name.
+- `REMOVABLE` will be set if the name is an existing instance variable.
 
-For all objects and keys:
+- `INVOCABLE` will be set if the object has a method of that name.
 
-`INVOCABLE` is never set, because currently `KEYS` does not include methods.
+- `INTERNAL` will be set if the name is an existing instance variable.
 
-`EXISTING` is set if `READABLE`, `MODIFIABLE`, `INVOCABLE`, `INTERNAL` or
-`REMOVABLE` are set.
+For all objects:
 
-`REMOVABLE` is always set to false.
+- `EXISTING` is set if `READABLE`, `MODIFIABLE`, `INVOCABLE`, `INTERNAL` or
+  `REMOVABLE` are set.
 
 ### `READ`
 
-The name must be an `int` (small Ruby `Fixnum`), or a Ruby `String` or `Symbol`,
-or a Java `String`.
-
-If the receiver is a Ruby `String` and the name is an integer, read a byte from
-the string, ignoring the encoding. If the index is out of range you'll get 0:
-
-`READ(string, integer)` → `string.get_byte(integer)`
-
-Otherwise, if the name starts with `@`, read it as an instance variable:
-
-`READ(object, "@name")` → `object.instance_variable_get("name")`
-
-Otherwise, if there is a method defined on the object with the same name as
-the name, perform a method call using the name as the called method name:
-
-`READ(object, name)` → `object.name` if `object.responds_to?(name)`
-
-Otherwise, if there isn't a method defined on the object with the same name as
-the name, and there is a method defined on the object called `[]`, call `[]`
-with the name as the argument:
-
-`READ(object, name)` → `object[name]` unless `object.responds_to?(name)`
+If there is a method defined on the object with the same name, return it as a
+(bound) `Method`.
 
-Otherwise throws `UnknownIdentifierException`.
+Otherwise, if there is a method called `[]` on the receiver, call
+`receiver[name]`.
 
-In all cases where a call is made no block is passed.
+Otherwise, look for an instance variable with the name and read it if one is
+found.
 
-An exception during a read will result in an `UnknownIdentifierException`.
+Otherwise, throw `UnknownIdentifierException`.
 
 ### `WRITE`
 
-The name must be a `String` or `Symbol`.
+If there is a method called `[]=` on the receiver, call
+`receiver[name] = value`.
 
-If the name starts with `@`, write it as an instance variable:
+Otherwise, write the name as an instance variable if the name is a legal
+instance variable name.
```

Comment:
Should we automatically prepend `@` to make it a legal variable name? If not, the caller really must know that it's working with Ruby.

Reply:
I would keep explicit `@`, then it's clear it's intended to be an instance variable write.

Reply:
No I was going to automatically add and remove the `@`. Is that not clear from the docs? Otherwise it's hard for other languages to `READ` and `WRITE` them.

Reply:
I think the docs are ambiguous around what constitutes a legal instance variable name. To the Ruby runtime, that means the leading `@` is already present.

Reply:
Sorry got confused. Hm indeed you cannot easily do `receiver.@ivar` in many languages. On the other hand other languages should need to write instance variables directly very rarely, so it might be fine let them have to use `receiver.instance_variable_set("@name", aValue)` or using interop API directly. I think no auto prepending is clearer.

Reply:
I've reread the doc, I would not have inferred that the `@` is prepended.

Reply:
For the record, I think we should prepend `@`. To me, using `instance_variable_get` and `instance_variable_set` is just an implementation detail of how we're choosing to respond to the interop calls. Otherwise, why not just use `INVOKE` for everything?

Reply:
If you think that instance variables shouldn't be used by other languages often then I don't think you really agree with the proposed changes.

The reason for reading instance variables is so that something like C can read a Ruby object as a struct.

Reply:
In javascript `r.f` is same as `r["f"]`, right? So I would expect that `r.f` gets me a bound method `f` if any and r["@f"] will get me a instance variable `@f`. The auto prepend may just confuse the situation and hide variables if they are named as the method (which might be more serious problem than just clarity, I did not realize before). In C a interop read function would have to be used (or instance_variable_get) but I think that's fine.

Reply:
> In C a interop read function would have to be used (or instance_variable_get) but I think that's fine.

But we can't (or don't want to) modify existing C code, that does things like `((RData *)object)->data`. That's a `READ` - that's what we get and we have to work with it.

And `r["@f"]` just looks bad to me - people are supposed to think they're getting an `x, y` object and be able to call `.x` on it without having to think about it's a Ruby object so append this funny symbol.

Reply:
Basically, I think `READ` and `WRITE` should have an implied `attr_accessor` that's being used.

Reply:
> Basically, I think READ and WRITE should have an implied attr_accessor that's being used.

That's how it is at the moment! And that's the bug I've got open to fix. Other people don't want a `READ` or `WRITE` executing arbitrary Ruby code in a method such as a `attr_accessor`.

Reply:
Here's another problem - do instance variables need to take precedence over methods? Because you don't want to try to read the instance variable and get the accessor instead.

Reply:
The RData use case is somewhat special case we are trying to create a Ruby code which will fit to a given C code. I think in that case we need to bend the Ruby object to fit the C without undermining the clarity of the interop message behavior. For this specific case I would create a misnamed instance variable `data` in the RData object or allow to define some kind of ivar alias.

I think usually users will write a new code to glue things together with interop, that's why I suggested that it's fine in C to call the read interop function to get an instance variable and I would like not to complicate it with auto-prepend.

Reply:
IMHO: `[]` always has precedence over reading methods/ivars if defined for READ.
`@` is an internal Ruby syntax detail, it doesn't make sense to me to expose it to the interop API. So imho we should prepend `@` automatically (@rschatz said the same for LLVM's `@` prefix).
I'd say state (ivars) have precedence over methods. Methods should really be used with INVOKE, which avoids the conflict and so in case of conflict we can still access the data (ivars).